rewrote statement sample section (#375)
Kattis/problem-package-format: spec/2023-07-draft.md

#### Samples Shown in the Problem Statement

By default, the `.in` and `.ans` pairs from `data/sample` are shown in the problem statement.

> If a `.out` file exists the `.out` file is shown instead of the `ans` file in the problem statement.

Author: i am not sure ^^'

#### Samples Shown in the Problem Statement

By default, the `.in` and `.ans` pairs from `data/sample` are shown in the problem statement.
This behavior can be customized by creating files with extension `.in.statement` and `.ans.statement`.
If one of these files exists, its contents replaces that of the file with the same name -- except the `.statement` extension -- for purposes of the problem statement.

> Note that `.statement` files are not validated.

Author: right now `.ans.statement` are validated so you can see them as `.out`

Standard in of the output validator corresponds to standard out of the submission, and standard out of the output validator corresponds to standard in of the submission.
Therefore, the output validator can control what information from the `.in` and `.ans` files is provided to the submission.

For interactive problems the `.in` and `.ans` pairs typically do not meaningfully capture how users are expected to interact with the output validator in such problems.
Therefore, it is advised to provide samples for the problem statement in one of the following two forms:
- An `.in.statement` and `.ans.statement` file as explained before.
- An `.interaction` file that contains lines starting with `<` and `>`,
  containing an interaction log with output from the output validator starting with `<` and output from the submission starting with `>`.

Note that you should choose one of these two options and not both.

If you want to make testing tools available for download, you can use [attachments](#attachments).

##### Multi-Pass Problems

> Multi-pass problems require a custom output validator, which interacts with the submission see [multi-pass](#multi-pass-validation).

Reviewer: So does interactive, why is this only written here?

Author: it is already written for interactive, i just added multi-pass which was missing

For multi-pass problems the `.in` and `.ans` pairs typically do not meaningfully capture how users are expected to interact with the output validator in such problems.
Therefore, it is advised to provide samples for the problem statement in one of the following two forms:
- An `.in.statement` and `.ans.statement` file as explained before.
- An `.interaction` file that contains lines starting with `<` and `>`, like for interactive problems.
  Passes are separated by a line containing `---` (three dashes).

>   When the problem is not interactive, simply start each pass by a number of lines starting with `<`, containing the sample input, followed by some lines starting with `>`, containing the sample answer.

Reviewer: I feel that the "Interactive Problems" and "Multi-Pass Problems" sections could definitely be merged, and probably also into the parent section?

Author: maybe but i am unsure how ^^'

Reviewer: I certainly does not have to be in this PR, in fact I think it make more sense to try to do that in a follow up. There is enough happening here...

> Note that you should choose one of these two options and not both.

Reviewer: Same as above.

> If you want to make testing tools available for download, you can use [attachments](#attachments).

Reviewer: Shouldn't this be in the "Samples Available for Download" section?

#### Samples Available for Download

> By default, the files in `data/sample` are available for download, except for `.interaction` files.

Reviewer: Not all. Definitely not `.png`, `.jpg`, `.jpeg`, `.svg`, I'm pretty sure not `.yaml` either.

> Note that, files with `.statement` replace those without the additional extension, e.g., `.in.statement` is renamed to `.in` and replaces it for the download.

Reviewer: Do we allow any other `.statement` files than `.in.statement` and `.and.statement`? I would say that the "Samples Shown in the Problem Statement" section above tells me "no", and this line implies "maybe yes?".

Author: I see no use case for this, but I also don't care if they exists?

Reviewer: Thinking a bit more about this. If we have problems with command line of file input, those would be given in `.yaml` and `.files` respectively, so they could need to be overridden for the statement.

> This behavior can be further customized by creating files with the additional extension `.download` like `.in.download`.

Reviewer: What `tc.*.download` files are allowed? Previously we said "any", but now we don't, and I would **not** interpret that as "any".

I would prefer if we could limit to only `.in` and `.ans`, but maybe that's not enough?

Reviewer: Sorry, it should probably be "...and `.files`", right?

If one of these files exists, its contents replaces that of the file with the same name -- except the `.download` extension -- for the problem download.
Note that `.download` files are not validated.

#### Validation

All `data/sample/*.in` files are input-validated.

> For non interavtive and non multi-pass problems all test cases in `data/sample` must be accepted by the output validator, with the `.ans` file used as both the `answer_file` and `team_output`.

Reviewer: ```suggestion
For non interactive and non multi-pass problems all test cases in `data/sample` must be accepted by the output validator, with the `.ans` file used as both the `answer_file` and `team_output`.
```

> If the preoblem is neither interactive nor multipass and `.in.statement` does not exits then `.ans.statement` is assumed to be a valid submission output for `.in`.

Reviewer: ```suggestion
If the problem is neither interactive nor multi-pass and `.in.statement` does not exits then `.ans.statement` is assumed to be a valid submission output for `.in`.
```

Reviewer: I think it would be better to day that `out` is always assumed to be valid submission output for `.in`. I.e. regardless of whether `.in.statement` exists. Otherwise there is no way to have a validated output when you want a statement input.

Then `.ans.statement` could be never validated. (And system can warn when you use this)

Author: updated this

Thus it has to pass output validation `output_validator .in .ans < .ans.statement`.

> All other files are not validated in any way.

Reviewer: I mean.... true... but we only ever validate `.in`, `.and`, `.out`, so it feels strange to say it.

Therefore, it is advised to instead provide samples for the problem statement in the form of a `.interaction` file.
This file contains lines starting with `<` and `>`, containing an interaction log with output from the output validator starting with `<` and output from the submission starting with `>`.

Note that if you are using a `.interaction` file you must not provide a `.statement` or `.out` file.

##### Multi-Pass Problems

Multi-pass problems require a custom output validator, which interacts with the submission see [multi-pass](#multi-pass-validation).

For multi-pass problems displaying two files is typically not meaningfully to capture how users are expected to interact with the output validator.
Therefore, it is advised to instead provide samples for the problem statement in the form of a `.interaction` file.
This file contains lines starting with `<` and `>`, like for interactive problems.
Passes are separated by a line containing `---` (three dashes).
When the problem is not interactive, simply start each pass by a number of lines starting with `<`, containing the sample input, followed by some lines starting with `>`, containing the sample output.

> Note that if you are using a `.interaction` file you must not provide a `.statement` or `.out` file.

Reviewer: ```suggestion
Note that if you are using a `.interaction` file you must not provide a `.in.statement`, `.ans.statement`, or `.out` file.
```

#### Samples Available for Download

> By default, the `.in` and `.ans` files in `data/sample` are available for download.

Reviewer: ```suggestion
By default, the `.in`, `.ans`, and `.files` files in `data/sample` are available for download.
```

Should we special case empty `.in` files? I.e. if there is no input, you currently must have an empty `.in` file (as opposed to a missing `.in` file, and I don't think we should change that). Should the empty `.in` file still be downloadable?

Author: both would be fine for me

Reviewer: @mzuenni How do you feel about adding mu suggestion here?

The other question could be broken out into a new issue.

Author: should this only be for empty `.in` files or do we just want to say that empty files are not included for downloads?

Reviewer: It's only needed for empty `.in` files or empty `.*.statement` and `.*.download` files. Any other file could be either not there, or overridden by an empty `.*.statement` or `.*.download`. We could even limit it so empty `.*.statement` and `.*.download` files, but then you would have to actually override empty `.in` files with empty `.in.statement` or `.in.download` files.

> Note that the content of `.in.statement` replaces that of `.in` and that the content of `.out` and `.ans.statement` replace that of `.ans` for the download.

Reviewer: ```suggestion
Note that the content of `.in.statement` replaces that of `.in` and that the content of `.out` or `.ans.statement` replace that of `.ans` for the download.
```

This behavior can be further customized by providing files with the extension `in.download` or `ans.download`.

> If one of these files exists, its contents replaces that of the file with the same name -- except the `.download` extension -- for the problem download.

Reviewer: Should we mention `.files.download`? I think "yes".

If you want to make other files -- like testing tools -- available for download, you can use [attachments](#attachments).

#### Validation

All `data/sample/*.in` files are input-validated.

> For non interactive and non multi-pass problems the `.out` files must pass the output validator.

Reviewer: Is it clear what that means?

Author: I would say yes.

> All other files are not validated in any way.

Reviewer: If there is no `.out` file (or `.interaction` or `.ans.statement`, so if the `.ans` is not overridden for the statement), then the `.ans` file must pass the output validator.
